compiler/qsc_eval/src/lib.rs

```diff
@@ -1369,19 +1418,67 @@ impl State {
         Ok(())
     }
 
-    fn update_array_index(
+    fn update_array_index_single(
         &self,
         env: &mut Env,
         globals: &impl NodeLookup,
         lhs: ExprId,
+        span: PackageSpan,
         index: usize,
         rhs: Value,
     ) -> Result<(), Error> {
         let lhs = globals.get_expr(self.package, lhs);
-        match (&lhs.kind, rhs) {
-            (&ExprKind::Var(Res::Local(node), _), rhs) => match env.get_mut(node) {
+        match &lhs.kind {
+            &ExprKind::Var(Res::Local(node), _) => match env.get_mut(node) {
                 Some(var) if var.is_mutable() => {
-                    var.value.update_array(index, rhs);
+                    var.value.update_array(index, rhs).map_err(|idx| {
+                        Error::IndexOutOfRange(idx.try_into().expect("index should be valid"), span)
+                    })?;
+                }
+                Some(_) => {
+                    unreachable!("update of mutable variable should be disallowed by compiler")
```

Comment:
```suggestion
                    unreachable!("update of immutable variable should be disallowed by compiler")
```

Unless I'm misreading something?

Reply:
good catch, typo from me!